### ISSUE TITLE
Optimize docker build

### ISSUE THREAD

**Issue by sataset** [NVIDIA/VideoProcessingFramework#355]
**Is your feature request related to a problem? Please describe.**
My main problem that building VPF requires a lot of tools and final builder image is quite big (~11 GiB).

**Describe the solution you'd like**
It would be perfect to see build guide for images like:
- `nvidia/cuda:11.7.0-base-ubuntu20.04`
- `python:3.8.13-slim`

**Describe alternatives you've considered**
So the main image I am using for my application is `nvcr.io/nvidia/tensorrt:22.05-py3`. However most of this image is useless as we are copying `.venv` and `poetry` folders which brings almost whole image due symlinks (?).

I tried building directly on that image without multi-staging and it resulted to 13 GiB image. The only pros to this approach was that I did not need to install any packages except av and several python-dev libs.

**Comment by sataset** [comment]
Half of the resulting image is vpf build requirements which are transferred with `.venv`:

![image](https://user-images.githubusercontent.com/28687456/179252124-954ffa0d-a894-413e-8d5c-dbaa838aead8.png)

**Comment by sataset** [comment]
I was able to successfully build VPF and include compiled library to the final image. Resulting size is 6.42GB.

I have attached `Dockerfile` and modified VPF install archive I used. I have removed Video Codec SDK due to license agreement, but I used version 11.1.5.

[Dockerfile.txt](https://github.com/NVIDIA/VideoProcessingFramework/files/9125079/Dockerfile.txt)
[vpf-no-torch.tgz](https://github.com/NVIDIA/VideoProcessingFramework/files/9125071/vpf-no-torch.tgz)

**Comment by rarzumanyan** [comment]
Hi @sataset 

Thanks a lot for your investigation and detailed report.
Docker support is a feature implemented by community members. I can't take this up (not an expert in this field).
Allow me some time, I'll ask our Docker people to give me a hand with this issue and I'll come back as I have a word from them.

**Comment by rarzumanyan** [comment]
>Maybe we can collaborate to make a single documentation page for docker builds explaining exactly libraries needed and compatible cuda, cudnn versions?

That would be great.
Having multiple Docker documents is my mistake. Since I didn't do anything Docker I also forgot to update instructions here and there.

>ffmpeg addition is unclear for me as it was not required with Dockerfile I provided. Did you check if libavformat-dev solves the issue for the https://github.com/NVIDIA/VideoProcessingFramework/issues/333?

#333 happens because FFMpeg has quickly changing API with some of its bits and pieces being deprecated between versions. At one moment of time FFMpeg version installed with package manager became API-incompatible with FFMpeg `master` ToT causing compilation issues.

IMO it would be better to have a single point of truth and simply build FFMpeg from source.

**Comment by rarzumanyan** [comment]
@sataset 

Can you please take a look at #357 ?

**Comment by sataset** [comment]
Left several [comments](https://github.com/NVIDIA/VideoProcessingFramework/issues/355#issuecomment-1187558502) on it for @gedoensmax. Will take a closer look on it this weekend.

**Comment by gedoensmax** [comment]
Hey @sataset, so ffmpeg addition is needed to build. `libavformat-dev` was already in this container but did not provide the correct needed header.
Feel free to open your own PR  i am happy to review it together with Roman. I did not look into building an as minimal as possible image using `cuda-base` or `python-slim`, but feel free to investigate in this direction. I have not seen the need to get deployment size optimized yet.

**Comment by sataset** [comment]
Sorry, did not have enough time previous weekends. Testing requires extra time as I have been using VPF `v1.1.1` and not `master`. Currently VPF release works with `apt` packages, and not with `FFmpeg:master`.
I have received an error during `Building CXX object PyNvCodec/TC/CMakeFiles/TC.dir/src/FfmpegSwDecoder.cpp.o`:
```
/opt/PyNvCodec/TC/src/FfmpegSwDecoder.cpp:57:3: error: 'AVCodecContext' does not name a type; did you mean 'VideoContext'?
```
which is similar to #243 I suppose.

For the provided build I have disabled torch extension compile, because I did not find the difference with installation from package manager.

## Questions
1. is it necessary to compile torch on build stage? Is it any different from installing it straight from `pip`?
2. is `FFmpeg:master` required only for `VPF:master` builds?

## Suggestions
Versions could be fixed correspondingly:
- `FFmpeg:master` for `VPF:master`
- `FFmpeg:nX.Y` for `VPF:X.Y.Z`

This is to avoid incompatibility issues that `FFmpeg:master` could raise.

FFmpeg repo have tags for specific versions, so fixing version should not be a big problem:
`git clone --depth 1 --branch n5.1 https://github.com/FFmpeg/FFmpeg.git`

From what I have experienced, developers tend to fix all versions for docker builds. Especially when local packages are used, e.g. CUDA, cuDNN, Video Codec SDK in our example. This way generated image will be static all the time which will guarantee its work.

**Comment by gedoensmax** [comment]
Hey could this be closed due to #370 ? 
Also about your questions - i am not sure what you mean by torch being build on the build stage. Furthermore your suggestion to fix versions on specific tags makes sense but i think VPF master has always required ffmpeg master right ? @rarzumanyan maybe you can help me out here.
